File: hw6-fs/consumer/src/Metrics.fs
module Warehouse.Metrics

open System
open System.Diagnostics
open Prometheus

let eventsProcessed =
    Metrics.CreateCounter(
        "events_processed_total",
        "Total events processed",
        CounterConfiguration(LabelNames = [| "event_type" |])
    )

let processingDuration =
    Metrics.CreateHistogram(
        "event_processing_duration_seconds",
        "Event processing duration",
        HistogramConfiguration(Buckets = [| 0.001; 0.005; 0.01; 0.05; 0.1; 0.5; 1.0 |])
    )

let consumerLag =
    Metrics.CreateGauge(
        "consumer_lag",
        "Consumer lag per partition",
        GaugeConfiguration(LabelNames = [| "partition" |])
    )

let cassandraWriteErrors =
    Metrics.CreateCounter("cassandra_write_errors_total", "Total Cassandra write errors")

let startTimer (histogram: Histogram) =
    let sw = Stopwatch.StartNew()

    { new IDisposable with
        member _.Dispose() =
            histogram.Observe(sw.Elapsed.TotalSeconds) }

File: hw6-kt/consumer/src/main/kotlin/warehouse/metrics/Metrics.kt
package warehouse.metrics

import io.prometheus.client.Counter
import io.prometheus.client.Gauge
import io.prometheus.client.Histogram

object Metrics {
    val consumerLag: Gauge = Gauge.build()
        .name("consumer_lag")
        .help("Consumer lag per partition")
        .labelNames("partition")
        .register()

    val eventsProcessed: Counter = Counter.build()
        .name("events_processed_total")
        .help("Total events processed")
        .labelNames("event_type")
        .register()

    val processingDuration: Histogram = Histogram.build()
        .name("event_processing_duration_seconds")
        .help("Event processing duration in seconds")
        .buckets(0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0)
        .register()

    val cassandraWriteErrors: Counter = Counter.build()
        .name("cassandra_write_errors_total")
        .help("Total Cassandra write errors")
        .register()
}
